arbnode/batch_poster.go

```diff
@@ -1532,11 +1556,72 @@ func (b *BatchPoster) maybePostSequencerBatch(ctx context.Context) (bool, error)
 			l1BoundMinTimestampWithBypass = arbmath.SaturatingUSub(timestampWithPadding, arbmath.BigToUintSaturating(maxTimeVariationDelaySeconds))
 		}
 	}
+	var lastPotentialMsg *arbostypes.MessageWithMetadata
+	var lastPotentialMsgPos arbutil.MessageIndex
+	var addMessageLoop func() bool
+	var getNextMessage func() (*arbostypes.MessageWithMetadata, error)
+	var breakLoopWhenErrorOccurs bool
 
-	for b.building.msgCount < msgCount {
-		msg, err := b.streamer.GetMessage(b.building.msgCount)
+	if b.espressoStreamer == nil {
```

Comment:
I think we could hide the code that sets most of the function closures here in another function so that we don't have the large set of nested if statements here in an already long function. That would also help with merge conflicts in this particularly sensitive part of the code. but I don't cleaning this up is super critical before merging as long as everyone understands it.